internal/reconstitution/writebuffer.go

```diff
@@ -129,44 +128,46 @@ func (w *writeBuffer) updateSlice(ctx context.Context, sliceNSN types.Namespaced
 		return false
 	}
 
-	// Allocate the status slice if needed
-	var status *apiv1.ResourceSliceStatus
-	var dirty bool
-	if len(slice.Status.Resources) != len(slice.Spec.Resources) {
-		dirty = true
-		status = slice.Status.DeepCopy()
-		status.Resources = make([]apiv1.ResourceState, len(slice.Spec.Resources))
+	// It's easier to pre-allocate the entire status slice before sending patches
+	// since the "replace" op requires an existing item.
+	if len(slice.Status.Resources) == 0 {
+		copy := slice.DeepCopy()
+		copy.Status.Resources = make([]apiv1.ResourceState, len(slice.Spec.Resources))
+		err = w.client.Status().Update(ctx, copy)
+		if err != nil {
+			logger.Error(err, "unable to update resource slice")
+			return false
+		}
+		slice = copy
 	}
 
-	// Use the newly allocated slice (if allocated), otherwise use the already allocated one
+	// Transform the set of patch funcs into a set of jsonpatch objects
 	unsafeSlice := slice.Status.Resources
-	if unsafeSlice == nil {
-		unsafeSlice = status.Resources
-	}
-
-	// The resource slice informer doesn't DeepCopy automatically for perf reasons.
-	// So we can apply the CheckFn to status pointers but not PatchFn.
-	// We'll deep copy the status here only when it needs to change.
+	var patches []*jsonPatch
 	for _, update := range updates {
 		unsafeStatusPtr := &unsafeSlice[update.SlicedResource.Index]
-		if update.CheckFn(unsafeStatusPtr) {
+		patch := update.PatchFn(unsafeStatusPtr)
```

Comment:
unsafe, very rusty of you :D

Reply:
Go did it first! 🤣 https://pkg.go.dev/unsafe